Codegen: python_torch_functions only include relevant operators

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #70021
* __->__ #68693

Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

peterbell10 committed:
Codegen: python_torch_functions only include relevant operators

Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Rebase to fix merge conflicts on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Fix mypy issues on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Commit to re-run CI on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Fix missing includes on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Rebase to fix merge conflicts on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

albanD commented:
@peterbell10 isn't this breaking the non per-op build? Can you try locally?
In particular, the ops/tensor.h and ops/from_blob.h tensors need to be handled in a specific way?
Do we want to make sure the OSS build works without per-op? Or I should just fix the internal build directly?

peterbell10 commented:
> @peterbell10 isn't this breaking the non per-op build? Can you try locally?

Works for me locally

> In particular, the `ops/tensor.h` and `ops/from_blob.h` tensors need to be handled in a specific way?

They are just regular files and not code generated, so no. You might have to add a new glob or something to the buck build so they get picked up.

> Do we want to make sure the OSS build works without per-op? Or I should just fix the internal build directly?

I would prefer the OSS build works with per operator headers. I've got a PR where I'm trying to add it to the CI. 

What errors are you seeing internally?

albanD commented:
> What errors are you seeing internally?

Only missing includes.
I though they would be missing from generated outputs. But they are indeed plain files. Let me add them to the buck specs and test again.
Thanks for the quick answer!

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Rebase and fix merge conflicts on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented on torch/csrc/api/include/torch/types.h:
```diff
@@ -7,6 +7,11 @@
 
+// TODO: These don't really belong here but torchvision builds in CI need them
+// Remove once the torchvision version being compiled in CI is updated
+#include <torch/library.h>
```
This include causes problems for other builds. In particular, it prevents including torch.h direclty from what I can see:

```
In file included from original_header.h:7:
In file included from torch/csrc/api/include/torch/torch.h:3:
In file included from torch/csrc/api/include/torch/all.h:8:
In file included from torch/csrc/api/include/torch/data.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader/stateful.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader/base.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader_options.h:4:
In file included from torch/csrc/api/include/torch/types.h:12:
xplat/caffe2/torch/library.h:397:39: error: no member named 'parseSchema' in namespace 'torch::jit'
  c10::FunctionSchema s = torch::jit::parseSchema(str);
                          ~~~~~~~~~~~~^
# More errors like that
```

Removing this include does allow the build to go smoothly.

peterbell10 replied:
Looks like the include of `function_schema_parser.h` was guarded but the uses of it weren't. The old code probably included it transitively via `variable.h` or something like that.

albanD replied:
Thanks for finding this out!

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

peterbell10 commented:
@pytorchbot ciflow rerun

peterbell10 committed:
Update on "Codegen: python_torch_functions only include relevant operators"


Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

[ghstack-poisoned]

peterbell10 commented:
Regarding the recent CI failures, I wasn't able to reproduce it until after doing `conda update --all` and it still fails in branches that don't include this PR, so I'm confident it's an unrelated dependency issue.
